// @vendure/admin-ui/esm2015/catalog/providers/routing/collection-resolver.js
import { Injectable } from '@angular/core';
import { Router } from '@angular/router';
import { BaseEntityResolver } from '@vendure/admin-ui/core';
import { getDefaultUiLanguage } from '@vendure/admin-ui/core';
import { DataService } from '@vendure/admin-ui/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
import * as i2 from "@vendure/admin-ui/core";
export class CollectionResolver extends BaseEntityResolver {
    constructor(router, dataService) {
        super(router, {
            __typename: 'Collection',
            id: '',
            createdAt: '',
            updatedAt: '',
            languageCode: getDefaultUiLanguage(),
            name: '',
            slug: '',
            isPrivate: false,
            description: '',
            featuredAsset: null,
            assets: [],
            translations: [],
            filters: [],
            parent: {},
            children: null,
        }, id => dataService.collection.getCollection(id).mapStream(data => data.collection));
    }
}
CollectionResolver.ɵprov = i0.ɵɵdefineInjectable({ factory: function CollectionResolver_Factory() { return new CollectionResolver(i0.ɵɵinject(i1.Router), i0.ɵɵinject(i2.DataService)); }, token: CollectionResolver, providedIn: "root" });
CollectionResolver.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
CollectionResolver.ctorParameters = () => [
    { type: Router },
    { type: DataService }
];
