// @vendure/admin-ui/esm2015/catalog/providers/routing/product-resolver.js
import { Injectable } from '@angular/core';
import { Router } from '@angular/router';
import { BaseEntityResolver } from '@vendure/admin-ui/core';
import { getDefaultUiLanguage } from '@vendure/admin-ui/core';
import { DataService } from '@vendure/admin-ui/core';
import * as i0 from "@angular/core";
import * as i1 from "@vendure/admin-ui/core";
import * as i2 from "@angular/router";
export class ProductResolver extends BaseEntityResolver {
    constructor(dataService, router) {
        super(router, {
            __typename: 'Product',
            id: '',
            createdAt: '',
            updatedAt: '',
            enabled: true,
            languageCode: getDefaultUiLanguage(),
            name: '',
            slug: '',
            featuredAsset: null,
            assets: [],
            description: '',
            translations: [],
            optionGroups: [],
            facetValues: [],
            variants: [],
            channels: [],
        }, (id) => dataService.product
            .getProduct(id)
            .refetchOnChannelChange()
            .mapStream((data) => data.product));
    }
}
ProductResolver.ɵprov = i0.ɵɵdefineInjectable({ factory: function ProductResolver_Factory() { return new ProductResolver(i0.ɵɵinject(i1.DataService), i0.ɵɵinject(i2.Router)); }, token: ProductResolver, providedIn: "root" });
ProductResolver.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
ProductResolver.ctorParameters = () => [
    { type: DataService },
    { type: Router }
];
