// @vendure/admin-ui/esm2015/order/components/payment-detail/payment-detail.component.js
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import { CurrencyCode } from '@vendure/admin-ui/core';
export class PaymentDetailComponent {
}
PaymentDetailComponent.decorators = [
    { type: Component, args: [{
                selector: 'vdr-payment-detail',
                template: "<vdr-labeled-data [label]=\"'order.payment-method' | translate\">\r\n    {{ payment.method }}\r\n</vdr-labeled-data>\r\n<vdr-labeled-data [label]=\"'order.amount' | translate\">\r\n    {{ payment.amount | localeCurrency: currencyCode }}\r\n</vdr-labeled-data>\r\n<vdr-labeled-data *ngIf=\"payment.errorMessage\" [label]=\"'order.error-message' | translate\">\r\n    {{ payment.errorMessage }}\r\n</vdr-labeled-data>\r\n<vdr-labeled-data *ngIf=\"payment.transactionId\" [label]=\"'order.transaction-id' | translate\">\r\n    {{ payment.transactionId }}\r\n</vdr-labeled-data>\r\n<vdr-labeled-data [label]=\"'order.payment-metadata' | translate\">\r\n    <vdr-object-tree [value]=\"payment.metadata\"></vdr-object-tree>\r\n</vdr-labeled-data>\r\n",
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: [""]
            },] }
];
PaymentDetailComponent.propDecorators = {
    payment: [{ type: Input }],
    currencyCode: [{ type: Input }]
};
